weave/trace/op.py

```diff
@@ -734,4 +785,294 @@ def as_op(fn: Callable) -> Op:
+        return self._context_value
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        if self.op.lifecycle_handler.has_finished:
+            raise OpCallError("Should not call finish more than once")
+
+        boxed_output = box.box(self.call.output)
+        client = weave_client_context.require_weave_client()
+        client.finish_call(self.call, boxed_output, exc_value, op=self.op)
+        # self.op.lifecycle_handler.has_finished = True
+
+        if not call_context.get_current_call():
+            print_call_link(self.call)
+        call_context.pop_call(self.call.id)
+
+        return await self.orig_contextmanager.__aexit__(exc_type, exc_value, traceback)
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        if self._aiterator is None:
+            raise TypeError("Context value is not async iterable")
+        try:
+            val = await self._aiterator.__anext__()
+        except StopAsyncIteration:
+            raise
+        self.op.lifecycle_handler.before_yield(self.call, val)
+        return val
+
+
+@contextmanager
+def _call_context(op: Op, call: Call, should_raise: bool):
+    """Contextmanager to safely handle errors and finish op calls.
+
+    This is used to wrap the execution of an op, and is used to handle errors and finish
+    the call.
+    """
+    try:
+        yield
+    except Exception as e:
+        exception = e
+        op.lifecycle_handler.after_error(call, exception)
+        if should_raise:
+            raise
+    else:
+        exception = None
+    finally:
+        if op.lifecycle_handler.has_finished:
+            raise OpCallError("Should not call finish more than once")
+        boxed_output = box.box(call.output)
+        op.lifecycle_handler.before_call_finish(call)
+        client = weave_client_context.require_weave_client()
+        client.finish_call(call, boxed_output, exception, op=op)
+
+        if not call_context.get_current_call():
+            print_call_link(call)
+
+        call_context.pop_call(call.id)
+
+
+def _wrap_generator(
+    op: Op, args: Any, kwargs: Any, *, call: Call, should_raise: bool
+) -> Callable:
+    func = op.resolve_fn
+
+    @wraps(func)
+    def _wrapped_sync_generator() -> Any:
+        with _call_context(op, call, should_raise):
+            for val in func(*args, **kwargs):
+                op.lifecycle_handler.before_yield(call, val)
+                yield val
+
+    return _wrapped_sync_generator()
+
+
+async def _wrap_async_generator(
+    op: Op, args: Any, kwargs: Any, *, call: Call, should_raise: bool
+) -> Callable:
+    func = op.resolve_fn
+
+    @wraps(func)
+    async def _wrapped_async_generator() -> Any:
+        with _call_context(op, call, should_raise):
+            async for val in await func(*args, **kwargs):
+                op.lifecycle_handler.before_yield(call, val)
+                yield val
+
+    return _wrapped_async_generator()
+
+
+async def _exc_op_async(
```

Comment:
This will be renamed to `_execute_op_async`.  It's named this way for now because it makes the diff easier to read.